lib/puppet/indirector/file_bucket_file/file.rb

```diff
@@ -148,15 +156,28 @@ def save_to_disk(bucket_file, files_original_path, contents_file, paths_file)
         # same way when the bytes are decoded as UTF-8, continue using system encoding
         Puppet::FileSystem.exclusive_open(paths_file, 0640, 'a+:external') do |f|
           if Puppet::FileSystem.exist?(contents_file)
-            verify_identical_file!(contents_file, bucket_file)
-            Puppet::FileSystem.touch(contents_file)
-          else
-            Puppet::FileSystem.open(contents_file, 0440, 'wb') do |of|
-              # PUP-1044 writes all of the contents
-              bucket_file.stream() do |src|
-                FileUtils.copy_stream(src, of)
-              end
+            if verify_identical_file(contents_file, bucket_file)
+              Puppet.info "FileBucket got a duplicate file #{bucket_file.checksum}"
+              Puppet::FileSystem.touch(contents_file)
+            elsif contents_file_matches_checksum?(contents_file, bucket_file.checksum_data, bucket_file.checksum_type)
+              # If the contents or sizes don't match, but the checksum does,
+              # then we've found a conflict (potential hash collision).
+              # Unlikely, but quite bad. Don't remove the file in case it's
+              # needed, but ask the user to validate.
+              # Note: Don't print the full path to the bucket file to avoid
+              # disclosing file system layout on server. Remove bucket_path from
+              # beginning of string.
+              raise Puppet::FileBucket::BucketError, _("Existing backup and new file have different content but same checksum, %{value}. Verify file '%{path}' inside the FileBucket and remove if incorrect.") %
+                { value: bucket_file.checksum, path: strip_bucket_path(contents_file, bucket_file) }
```

Comment:
What about logging a message with the full path using `Puppet.error` (so we're not disclosing any file info) and raise an exception with only the `sum`, so that we don't need to `strip_bucket_path`?